### PR TITLE
Sync Service/Deployment changes

### PR DESCRIPTION
**Pull request by runakash**
*Issue #, if available:* Fixes https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/20

*Description of changes:* Adding the listener to the Service Export Controller to track EndpointSlice events. EndpointSlice events are triggered by Service or Deployment changes. Refactor ServiceExport to handle the deletion scenario when the Service corresponding to ServiceExport is deleted. Also, updating the e2e test to accommodate the scale-up scenario. Improvements in unit tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/controllers/serviceexport_controller.go [hunk 1]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+	service := v1.Service{}
+	namespacedName := types.NamespacedName{Namespace: serviceExport.Namespace, Name: serviceExport.Name}
+	if err := r.Client.Get(ctx, namespacedName, &service); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Error(err, "no Service found for ServiceExport",
+				"Namespace", serviceExport.Namespace, "Name", serviceExport.Name)
+			// Mark ServiceExport to be deleted, if the corresponding Service is not found
+			isServiceExportMarkedForDelete = true
+		} else {
+			r.Log.Error(err, "error with fetching service",
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:89]: nit: "with" unnecessary.

File: pkg/controllers/serviceexport_controller.go [hunk 2]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
-	if !controllerutil.ContainsFinalizer(svcExport, serviceExportFinalizer) {
-		controllerutil.AddFinalizer(svcExport, serviceExportFinalizer)
-		if err := r.Update(ctx, svcExport); err != nil {
+func (r *ServiceExportReconciler) handleUpdate(ctx context.Context, serviceExport *v1alpha1.ServiceExport, service *v1.Service) (ctrl.Result, error) {
+
+	// Add the finalizer to the service export if not present, ensures the ServiceExport won't be deleted
+	if !controllerutil.ContainsFinalizer(serviceExport, ServiceExportFinalizer) {
+		controllerutil.AddFinalizer(serviceExport, ServiceExportFinalizer)
+		if err := r.Update(ctx, serviceExport); err != nil {
+			r.Log.Error(err, "error with adding finalizer",
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:109]: nit: "with" unnecessary.

File: pkg/controllers/serviceexport_controller.go [hunk 3]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
-		log.Error(err, "no service found for ServiceExport",
-			"Namespace", svcExport.GetNamespace(), "Name", svcExport.Name)
-		return ctrl.Result{}, client.IgnoreNotFound(err)
-	}
-
-	endpoints, err := r.extractEndpoints(ctx, &svc)
+	r.Log.Info("updating Cloud Map service", "namespace", service.Namespace, "name", service.Name)
+	cmService, err := r.createOrGetCloudMapService(ctx, service)
 	if err != nil {
+		r.Log.Error(err, "error with fetching service from Cloud Map",
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:118]: with nit

File: pkg/controllers/serviceexport_controller.go [hunk 4]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
-	changes := model.Changes{
-		Create: endpoints,
-	}
-
-	log.Info("updating Cloud Map service", "namespace", svc.Namespace, "name", svc.Name)
-	srv, err := r.Cloudmap.GetService(ctx, svc.Namespace, svc.Name)
+	endpoints, err := r.extractEndpoints(ctx, service)
 	if err != nil {
-		log.Error(err, "error when fetching service from Cloud Map API", "namespace", svc.Namespace, "name", svc.Name)
+		r.Log.Error(err, "error with extracting endpoints",
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:125]: with nit

File: pkg/controllers/serviceexport_controller.go [hunk 5]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
 		// merge creates and updates (Cloud Map RegisterEndpoints can handle both)
-		upserts := append(changes.Create, changes.Update...)
+		upserts := changes.Create
+		upserts = append(upserts, changes.Update...)
 
-		if err := r.Cloudmap.RegisterEndpoints(ctx, svc.Namespace, svc.Name, upserts); err != nil {
-			log.Error(err, "error when registering endpoints to Cloud Map",
-				"namespace", svc.Namespace, "name", svc.Name)
+		if err := r.CloudMap.RegisterEndpoints(ctx, service.Namespace, service.Name, upserts); err != nil {
+			r.Log.Error(err, "error with registering endpoints to Cloud Map",
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:143]: with nit

File: pkg/controllers/serviceexport_controller.go [hunk 6]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+				"namespace", service.Namespace, "name", service.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
-	if deleteRequired {
-		if err := r.Cloudmap.DeleteEndpoints(ctx, svc.Namespace, svc.Name, changes.Delete); err != nil {
-			log.Error(err, "error when deleting endpoints from Cloud Map",
-				"namespace", srv.Namespace, "name", srv.Name)
+	if changes.IsDeleted() {
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:149]: IsDeleted sounds like changes have been deleted. HasDeletes?

File: pkg/controllers/serviceexport_controller.go [hunk 7]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		if err := r.CloudMap.DeleteEndpoints(ctx, service.Namespace, service.Name, changes.Delete); err != nil {
+			r.Log.Error(err, "error when deleting endpoints from Cloud Map",
+				"namespace", cmService.Namespace, "name", cmService.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
-	if !createRequired && !updateRequired && !deleteRequired {
-		log.Info("no changes to export", "namespace", svc.Namespace, "name", svc.Name)
+	if changes.IsNone() {
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:157]: IsEmpty()?

File: pkg/controllers/serviceexport_controller.go [hunk 8]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
-	if controllerutil.ContainsFinalizer(svcExport, serviceExportFinalizer) {
+func (r *ServiceExportReconciler) createOrGetCloudMapService(ctx context.Context, service *v1.Service) (*model.Service, error) {
+	cmService, err := r.CloudMap.GetService(ctx, service.Namespace, service.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if cmService == nil {
+		if err := r.CloudMap.CreateService(ctx, service.Namespace, service.Name); err != nil {
+			r.Log.Error(err, "error with creating a new service in Cloud Map",
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:172]: with nit

File: pkg/controllers/serviceexport_controller.go [hunk 9]
```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
 		}
-		if srv != nil {
-			if err := r.Cloudmap.DeleteEndpoints(ctx, srv.Namespace, srv.Name, srv.Endpoints); err != nil {
-				log.Error(err, "error when deleting endpoints from Cloud Map",
-					"namespace", srv.Namespace, "name", srv.Name)
+		if cmService != nil {
+			if err := r.CloudMap.DeleteEndpoints(ctx, cmService.Namespace, cmService.Name, cmService.Endpoints); err != nil {
+				r.Log.Error(err, "error while deleting endpoints from Cloud Map",
+					"namespace", cmService.Namespace, "name", cmService.Name)
 				return ctrl.Result{}, err
```
Comment (astaticvoid) [inline pkg/controllers/serviceexport_controller.go:199]: I guess in the future we'll return something other than empty struct.

**Review by astaticvoid — APPROVED** [review]

**Review by astaticvoid — APPROVED** [review]

File: pkg/controllers/serviceexport_controller.go (hunk 9 above)
Comment (runakash) [reply pkg/controllers/serviceexport_controller.go:199]: Right now our strategy seems to be just log error, and move on. We have to probably figure out fault-tolerance strategy before GA.

**Review by bendu — APPROVED** [review]

**Pull request merged**